### DISCUSSION TITLE
Dynamic lift and wait times

### DISCUSSION THREAD

**Discussion by msspad** [sn4k3/UVtools#291] [Q&A] [answered]
UVTools is a serious work horse for me, thanks for all your work! Some questions & remarks:
1) https://github.com/sn4k3/UVtools/wiki/Rest-times-and-TSMC#how-to-tune-acceleration Is there a way to see the default configuration (acceleration, jerk, etc) of a Phrozen Sonic Mini 4k (or any Chitubox machine besides elegoo code)?
2) Dynamic lift: I thought it can make sense to set a lower lift speed for a higher lift height (because more mass should be pulled with lower speed (or even better acceleration)? a warning prevents this currently
3) The format of "Wait times" 2/0.5/0 is: 2 seconds wait before cure / 0.5 seconds wait after cure / 0 seconds wait after lift. correct?
4) "Edit print parameters" with ctb v4 changes 4 parameters even without touching anything:
![image](https://user-images.githubusercontent.com/6191764/133060455-c8a55f34-e40a-4a77-82ab-fa066a373553.png)

**Comment by sn4k3** [comment, marked as answer]
1. It should be possible with a serial link, but not that easy... See: https://drive.google.com/file/d/18Ud_x-aR85m5qi4uiX9_ze26UC_jC4kA/view?fbclid=IwAR0tLzLQnCN4j1EHJK0zeDOEmWeYW_sROovdGYMQllT6UdmvzG8IVrceZw8
2. _Make sense to set a lower lift speed for a higher lift height:_ That is how it work already, if you read description it will tell you that.
3. Correct, highlight labels show tooltip with meaning
4. Because i dont allow 0 as value, it's normal thing. Nothing wrong, no side effects

**Reply by msspad** [reply]
1. very interesting, thank you! So you recommend reducing max jerk from 2 to 0.4 mm/s3 and the max acceleration 800 to 400m/s2 (for the SM4k), I will experiment with those values.
2. mmm, I'm confused. Let's say Minimum (mass) is 2mm/50mm/min and Maximum (mass) is 4mm/100m/min. Then I would expect a layer with lift height 2mm to have 100m/s lift speed? But for layers with 2mm lift height, I find anything between 50 to 72m/min for lift speed?
3. true, I didn't see the tooltip in the bottom bar. I tried it in the "Layer data" table where no tooltips are present.
4.  sure, no big thing, just somewhat surprising when the user does not change anything but the software suggests saving changed values.

**Reply by sn4k3** [reply]
1. It's a start point, you need to test and see what goes best with you
2. Lift minimum speed is linked to largest height/mass, and maximum speed is linked to lower mass. Forget everything, you just need to setup your min/max lift height and speed. UVtools does the rest, the field is below the other doesnt mean it will correlate with it. Probally i could give other layout to it...

**Reply by msspad** [reply]
ok, thanks for your explanation. will look it up in the code, just out of interest.
keep up your great work!

**Reply by sn4k3** [reply]
Here: https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Operations/OperationDynamicLifts.cs#L317

**Reply by msspad** [reply]
I love open source tools :-)
_s = layer.NonZeroPixelCount / maxNormalPixels_
UVTools scales maxLiftSpeed with s. minLiftSpeed is only a security threshold but it's not linked to _s_ or _maxNormalPixels_. That's totally fine and it might result in better printing results.

What I expected was that you scale the range, like 
_r = maxLiftSpeed - minLiftSpeed
liftSpeed = maxLiftSpeed - (r * (layer.NonZeroPixelCount - minNormalPixels) / maxNormalPixels)_
But as said before, I have no indication that this would give better printing result.

**Reply by sn4k3** [reply]
I could add that variation if you wish, and let user decide which approach to take, still min and max would be clamp

**Reply by msspad** [reply]
I have no facts to show that it works better, it's only more intuitive for me to understand. I would use this option if I had a choice.
Clamping: the formula squeezes all values into the range of [minLiftSpeed, maxLiftSpeed]. However, if any variable is non-integer, clamping is still necessary for numeric imprecision.

**Reply by sn4k3** [reply]
Ok i can give the option.
Just tell me what that options names can be called (to show on GUI) what can i call them?

**Reply by msspad** [reply]
not sure, how about "full range"?
description "squeezes lift height and lift speed within full range of min/max values. E.g. the layer with the least pixels gets minLiftHeight and maxLiftSpeed, the layer with the most pixels gets maxLiftHeight and minLiftSpeed."

**Reply by sn4k3** [reply]
Need to be more specific, we will have two methods.

Method1 = My
Method2 = Your

I need name and description for both. Can you purpose something?

Also give the math for LiftHeight, you only post for Speed

**Comment by msspad** [comment]
Method1 = traditional
Method2 = full range

The correct formula, by the way:
pixel_ratio = (layer.NonZeroPixelCount - minNormalPixels) / (maxNormalPixels - minNormalPixels);  // pixel_ratio is between 0 and 1
liftHeight = (_minLiftHeight + ((_maxLiftHeight - _minLiftHeight) * pixel_ratio)).Clamp(_minLiftHeight, _maxLiftHeight);
liftSpeed = (_maxLiftSpeed - ((_maxLiftSpeed - _minLiftSpeed) * pixel_ratio).Clamp(_minLiftSpeed, _maxLiftSpeed);

**Reply by sn4k3** [reply]
Ok will also need a short text description for traditional and full range, or i will have people asking what that does...

**Reply by msspad** [reply]
traditional: reduces maximal lift height with the number of pixels in layer divided by maximal number of pixels in any layer. Increases the minimal speed with the same ratio.
full range: squeezes lift height and lift speed within full range of min/max values. E.g. the layer with the least pixels gets minimal lift height and maximal lift speed. The layer with the most pixels gets maximal lift height and minimal lift speed.

**Reply by sn4k3** [reply]
Thanks, it's already on the repo

**Reply by sn4k3** [reply]
I tested your math, but it dont look very good...
It drastic change lift height and speed with a minor change, don't look correct to have this.
https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Operations/OperationDynamicLifts.cs#L345
Check this almost identical layers:

![image](https://user-images.githubusercontent.com/113281/133544202-a4d0fc19-e99e-4ab7-9101-5281c4718ead.png)
![image](https://user-images.githubusercontent.com/113281/133544212-9074d1c9-3f6a-4a19-a8d6-a9f836a44cb7.png)
![image](https://user-images.githubusercontent.com/113281/133544220-82d0c042-c3dc-44fa-b009-f1cdaf59d1ba.png)
![image](https://user-images.githubusercontent.com/113281/133544226-2e1729b9-6d70-4c77-a9fc-3d364d2a033e.png)

**Reply by msspad** [reply]
The idea behind Dynamic lift is to speed up printing while maintaining quality. There are typically only very few bottom layers and you won't gain any noticeable time saving with optimizing bottom layers.
In your example, it seems minLiftSpeed=50mm/min, maxLiftSpeed=150mm/min. When you have 4 bottom layers, both those values will be assigned to a layer. The great thing about "full range" is that you know exactly the range of applied liftSpeed and liftHeight. So when you have 4 layers and they do have about the same number of pixels, why should you vary the liftSpeed by 100mm/min?

**Reply by sn4k3** [reply]
That was a test to see how it behave, i agree that bottoms should not be optimized, but still i have to test how it does. The minor change in pixels looks so small to produce such values and we have to account every usage, good or bad and protect user from himself.
I tested on normal layers and looked better

**Reply by sn4k3** [reply]
I also have revised variable names to be less confusing. Please review the math and confirm me if is ok: https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Operations/OperationDynamicLifts.cs#L345

**Reply by msspad** [reply]
sure, testing makes a lot of sense - that's what great software designers do :-) Also, when you open the dynamic lift window, only normal layers are selected (what is great) and selecting bottom layers needs to be made with intention (when opening the dropdown).
new variable names and math look fine to me.

**Reply by sn4k3** [reply]
Ok, here the new aspect 
![UVtools_2021-09-16_18-55-50](https://user-images.githubusercontent.com/113281/133661396-dedebfff-48c6-43d3-9193-e74364fc2b19.png)

**Reply by msspad** [reply]
looks great!
1) how about replacing "mass" in first 2 lines of the description, e.g. 
"Generate dynamic lift height and speeds for each layer relative to its number of pixels. More pixels in a layer require higher lift height and lower speed while fewer pixels can go with smaller lift height and higher speed."
2) last line, I guess it technically a challenge to add a link to "[few printers support](https://github.com/sn4k3/UVtools/wiki/Printer-compability-with-per-layer-settings-and-advanced-tools)"?

**Reply by sn4k3** [reply]
> 1. how about replacing "mass" in first 2 lines of the description, e.g.
>    "Generate dynamic lift height and speeds for each layer relative to its number of pixels. More pixels in a layer require higher lift height and lower speed while fewer pixels can go with smaller lift height and higher speed."

Technically pixels are the mass, the more pixels you cure the more the mass you have. I think mass term is more easy to understand for users?

> 2. last line, I guess it technically a challenge to add a link to "[few printers support](https://github.com/sn4k3/UVtools/wiki/Printer-compability-with-per-layer-settings-and-advanced-tools)"?

Correct, had to change a lot just to able to insert links (Richtext)

**Reply by msspad** [reply]
yes, mass is also fine but pixels is already often used in UVTools and it also appears in the dropdown "...number of pixels...". You can put mass in the dropdown, then it's at least consistent in the window?

**Reply by sn4k3** [reply]
WIll update accordingly
